src/core/thread/network_diagnostic_tlvs.hpp

```diff
@@ -0,0 +1,1307 @@
+     * @retval FALSE  If the TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(NetworkDiagnosticTlv); }
+
+    /**
+     * This method returns the Status value.
+     *
+     * @returns The Status value.
+     *
+     */
+    uint16_t GetSupplyVoltage(void) const { return mSupplyVoltage; }
+
+    /**
+     * This method sets the SupplyVoltage value.
+     *
+     * @param[in]  aSupplyVoltage  The SupplyVoltage value.
+     *
+     */
+    void SetSupplyVoltage(uint16_t aSupplyVoltage) { mSupplyVoltage = aSupplyVoltage; }
+
+private:
+    uint16_t mSupplyVoltage;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Child Table Entry generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class ChildTableEntry
+{
+public:
+    /**
+     * This method returns the Version value.
+     *
+     * @returns The Version value.
+     *
+     */
+    uint8_t GetTimeout(void) const { return mTimeout; }
+
+    /**
+     * This method sets the Version value.
+     *
+     * @param[in]  aVersion  The Version value.
+     *
+     */
+    void SetTimeout(uint8_t aTimeout) { mTimeout = aTimeout; }
+
+    /**
+     * This method returns the Version value.
+     *
+     * @returns The Version value.
+     *
+     */
+    uint16_t GetChildId(void) const { return mChildId; }
+
+    /**
+     * This method sets the Version value.
+     *
+     * @param[in]  aVersion  The Version value.
+     *
+     */
+    void SetChildId(uint16_t aChildId) { mChildId = aChildId; }
+
+    /**
+     * This method returns the Version value.
+     *
+     * @returns The Version value.
+     *
+     */
+    uint8_t GetMode(void) const { return mMode; }
+
+    /**
+     * This method sets the Version value.
+     *
+     * @param[in]  aVersion  The Version value.
+     *
+     */
+    void SetMode(uint8_t aMode) { mMode = aMode; }
+
+
+private:
+    uint8_t mTimeout: 5;
+    uint8_t mReserved: 2;
+    uint16_t mChildId: 9;
+    uint8_t mMode: 8;
+} OT_TOOL_PACKED_END;
+
+
+/**
+ * This class implements Child Table TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class ChildTableTlv: public NetworkDiagnosticTlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void) { SetType(kChildTable); SetLength(sizeof(*this) - sizeof(NetworkDiagnosticTlv)); }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval TRUE   If the TLV appears to be well-formed.
+     * @retval FALSE  If the TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(NetworkDiagnosticTlv); }
+
+    uint8_t GetNumEntries() const { return GetLength() / sizeof(ChildTableEntry); }
+
+    ChildTableEntry &GetEntry(uint8_t i) {
+        return mChildTableEntry[i];
+    }
+
+private:
+    ChildTableEntry mChildTableEntry[Thread::Mle::kMaxChildren];
```

Comment:
Probably better to iterate through each child entry individually, rather than trying to buffer everything here.